### DISCUSSION TITLE
`statement.compile`: how get the `:id` variable from `:id::INT`

### DISCUSSION THREAD

**Discussion by Fullchee** [sqlalchemy/sqlalchemy#8953] [Usage Questions] [answered]
How can I get SQLAlchemy's statement.compile to interpret `:id::INT` the same way that it interprets`:id ::INT`?

I'm getting this issue with versions 1.4.44 and '2.0.0b4'

```python
from sqlalchemy import text
from sqlalchemy.dialects import postgresql
```

If I add whitespace around `:id`, it works as expected

```python
>>> str(text('select :id ::INT').compile(dialect=postgresql.dialect()))
'select %(id)s ::INT'
```

However, if I don't have any whitespace like `:id::INT`, then it doesn't detect `:id` as a variable

```python
>>> str(text('select :id::INT').compile(dialect=postgresql.dialect()))
'select :id::INT'
```

I eventually want to run

```python
from django.db import connection
from sqlalchemy import text
from sqlalchemy.dialects import postgresql

with connection.cursor() as cursor:
    cursor.execute(str(text(some_sql_string).compile(dialect=postgresql.dialect())))
```

where I don't own the `some_sql_string` so I can't make changes to it

**Comment by zzzeek** [comment, marked as answer]
well easiest way is to use cast instead:

```py
>>> str(text('select cast(:id AS INT)').compile(dialect=postgresql.dialect()))
'select cast(%(id)s AS INT)'
```

however, when using `text()`, [colons can be rendered by escaping them](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.text)

```py
>>> str(text('select :id\:\:INT').compile(dialect=postgresql.dialect()))
'select %(id)s::INT'
```

**Reply by Fullchee** [reply]
Is there a parameter I can pass into `.compile` so that I don't have to modify the string but SQLAlchemy can still understand that `:id` is the variable?

I'm reading some SQL strings that I don't have ownership over so I can't modify them

If I could modify them, I would just add whitespace around the variables and it works that way

I'll update the question to be more clear

**Reply by Fullchee** [reply]
btw ty for the super quick reply!!

**Reply by zzzeek** [reply]
do you know the parameter name is "id" to start with?  how do you know the SQL string doesnt have a parameter named "INT" ?  (the answer to that is how we will decide how you can parse this SQL)

**Reply by zzzeek** [reply]
let me answer my own question on this - replace all "::" with "\:\:":

```py
>>> str(text('select :id::INT'.replace("::", "\:\:")).compile(dialect=postgresql.dialect()))
'select %(id)s::INT'
```

**Reply by Fullchee** [reply]
Yeah, I should be able to get a list of variables (from a regex)

and then I can just wrap those variables in parentheses

**Reply by Fullchee** [reply]
But is there a reason why SQLAlchemy doesn't handle this case?

When I run a query in PyCharn, it's able to handle that case

**Reply by zzzeek** [reply]
it would be very difficult to use an escaping format where only odd numbers of colons need to be escaped

such as

`SELECT * FROM table WHERE ident = 'FirstCode:::SecondCode'`

we try to stay away from things that are very surprising like that

**Reply by Fullchee** [reply]
@zzzeek I've found a edge case that where replacing `::` with `\:\:` doesn't work 😞

```python
from sqlalchemy import text
from sqlalchemy.dialects import postgresql
str(text("select * from table where id = :id::INT and column = ' :id::INT '".replace('::', '\:\:')).compile(dialect=postgresql.dialect()))
```

**Reply by zzzeek** [reply]
> When I run a query in PyCharn, it's able to handle that case

I would assume you have a PostgreSQL console in your pycharm.  SQLAlchemy can't replicate the PostgreSQL statement parser

**Reply by zzzeek** [reply]
> @zzzeek I've found a edge case that where replacing `::` with `\:\:` doesn't work disappointed

well that would be a case that SQLAlchemy couldn't do either then :)

any reason you dont want to just have real PostgreSQL SQL strings and pass them to your driver?

**Reply by Fullchee** [reply]
What are real PostgreSQL SQL strings? 😮

**Reply by Fullchee** [reply]
My goal is to

1. Replace SQL variables in the SQL string with their value
2. Avoid SQL injection

**Reply by CaselIT** [reply]
I think by "real PostgreSQL SQL strings" mike meant strings that are understood directly by the driver in use, so with sqlalchemy you would use `exec_drover_sql(sql_str, params)` directly without doing any change to the string at all.

These would depend on the driver in use, for example for psycopg you would use pyformat `%(name)s`, asyncpg (in 1.4) qmarks `?` (on v2 you can use the native numeric dollar `$1`)

**Reply by Fullchee** [reply]
I see!

So `psycopg2` (the underlying driver) would prevent SQL injection?

**Reply by CaselIT** [reply]
from psycopg2 point of view there is no difference between
```py
conn= engine.connect()
# use execute
conn.execute(sa.text('select :value'), {'value': 1}).scalar()
# use exec_driver_sql
conn.exec_driver_sql('select %(value)s', {'value': 1}).scalar()
```
in both cases it sees `select %(value)s` with parameters `{'value': 1}`
